login/service_blackbox_test.go

```diff
@@ -690,3 +691,156 @@ func (c *dummyOauth2Config) Exchange(ctx netcontext.Context, code string) (*oaut
 	token = token.WithExtra(extra)
 	return token, nil
 }
+
+func (s *serviceBlackBoxTest) TestKeycloakAuthorizationRedirectForAuthorize() {
+	rw := httptest.NewRecorder()
+	u := &url.URL{
+		Path: fmt.Sprintf("/api/authorize"),
+	}
+	req, err := http.NewRequest("GET", u.String(), nil)
+	if err != nil {
+		panic("invalid test " + err.Error()) // bug
+	}
+
+	// The user clicks login while on ALM UI.
+	// Therefore the referer would be an ALM URL.
+	refererUrl := "https://alm-url.example.org/path"
+	req.Header.Add("referer", refererUrl)
+
+	prms := url.Values{}
+
+	prms.Add("response_type", "code")
+	prms.Add("redirect_uri", "https://openshift.io/somepath")
+	prms.Add("client_id", "5dec5fdb-09e3-4453-b73f-5c828832b28e")
+	prms.Add("state", uuid.NewV4().String())
+
+	ctx := context.Background()
+	goaCtx := goa.NewContext(goa.WithAction(ctx, "AuthorizeTest"), rw, req, prms)
+	authorizeCtx, err := app.NewAuthorizeAuthorizeContext(goaCtx, req, goa.New("LoginService"))
+	if err != nil {
+		panic("invalid test data " + err.Error()) // bug
+	}
+	require.Nil(s.T(), err)
+
+	err = s.loginService.PerformAuthorize(authorizeCtx, s.oauth, s.Configuration)
+
+	assert.Equal(s.T(), 307, rw.Code)
+	assert.Contains(s.T(), rw.Header().Get("Location"), s.oauth.Endpoint.AuthURL)
+	assert.NotEqual(s.T(), rw.Header().Get("Location"), "")
+}
+
+func (s *serviceBlackBoxTest) TestValidOAuthAuthorizationCodeForAuthorize() {
+
+	_, callbackCtx := s.AuthorizeCallback("valid_code", make(map[string]string))
+	_, err := s.loginService.VerifyState(callbackCtx, callbackCtx.State.String(), callbackCtx.Code)
+	require.Nil(s.T(), err)
+
+	keycloakToken, err := s.loginService.GetTokenFromAuthorizationCode(callbackCtx, callbackCtx.Code, s.dummyOauth)
+	require.Nil(s.T(), err)
+	require.NotNil(s.T(), keycloakToken)
+	require.Nil(s.T(), err)
+}
+
+func (s *serviceBlackBoxTest) TestInvalidOAuthAuthorizationCodeForAuthorize() {
+
+	rw, callbackCtx := s.AuthorizeCallback("invalid_code", make(map[string]string))
+	_, err := s.loginService.VerifyState(callbackCtx, callbackCtx.State.String(), callbackCtx.Code)
+	require.Nil(s.T(), err)
+
+	keycloakToken, err := s.loginService.GetTokenFromAuthorizationCode(callbackCtx, "INVALID_OAUTH2.0_CODE", s.oauth)
+	require.NotNil(s.T(), err)
+	require.Nil(s.T(), keycloakToken)
+
+	assert.Equal(s.T(), 401, rw.Code)
+}
+
+func (s *serviceBlackBoxTest) TestInvalidOAuthStateForAuthorize() {
+
+	rw, callbackCtx := s.AuthorizeCallback("invalid_state", make(map[string]string))
+	_, err := s.loginService.VerifyState(callbackCtx, callbackCtx.State.String(), callbackCtx.Code)
+	require.NotNil(s.T(), err)
+	assert.Equal(s.T(), 401, rw.Code)
+}
+
+func (s *serviceBlackBoxTest) AuthorizeCallback(testType string, extraParams map[string]string) (*httptest.ResponseRecorder, *app.CallbackAuthorizeContext) {
```

Comment:
This method doesn't need to be exposed outside the package.

Reply:
https://github.com/fabric8-services/fabric8-auth/pull/210/commits/24b633c4334765f4623bff85a089c206bae145b8